### PR TITLE
Feature: mark item done

### PR DESCRIPTION
**Pull request by agonzalez0515**

### CODE REVIEW

File: src/main/java/webserver/controllers/AppController.java [hunk 1]
```diff
@@ -27,10 +33,15 @@
...
+    public static Callback<Request, String> getDirectoryFile = (request) -> {
+        String body = ResponseBody.getHtml(request.getPath());
+        return getResponseBuilder(200, body);
+    };
+
     private static String getResponseBuilder(int statusCode, String body) {
         return new Response.Builder(statusCode)
-                    .withHeader("Content-Type", "text/html; charset=utf-8")
-                    .withHeader("Content-Length", Integer.toString(body.length()))
+                    .withHeader(HTTP.CONTENT_TYPE, "text/html; charset=utf-8")
```
Comment (tbaik) [inline src/main/java/webserver/controllers/AppController.java]: let's rename HTTP to be HTTP_HEADERS if it's just for headers!

File: src/main/java/webserver/controllers/TodoController.java [hunk 2]
```diff
@@ -13,148 +13,128 @@
...
 
         } else {
-            
-            String title = getProperty(todoItem, "title");
-            String text = getProperty(todoItem, "text");
-            int todoId = Integer.parseInt(getProperty(todoItem, "id"));
-
-            Todo td = new Todo(title, text, todoId);
-            html = MustacheUtil.executeTemplate(template, td);
+            Todo todo = Todo.fromJson(todoItem);
```
[the remark this answers is not in the document]
Comment (tbaik) [reply src/main/java/webserver/controllers/TodoController.java]: ^wrote this one prior to our meeting. I think you have an idea of what to do!

File: src/main/resources/todos.mustache [hunk 3]
```diff
@@ -36,4 +36,29 @@
             &copy 2019 @angieg0nzalez | 8th Light Apprenticeship
         </footer>
     </body>
+    <script>
+        document
+        .getElementById("todo-form")
+        .addEventListener("submit", (event) => {
+        event.preventDefault();
+        var formData = "";
```
Comment (tbaik) [inline src/main/resources/todos.mustache]: you should be able to use const now I think!

File: src/main/java/webserver/request/Parser.java [hunk 4]
```diff
@@ -38,6 +41,20 @@ public Request parse(BufferedReader in) throws IOException {
                 .build();
     }
 
+    public static HashMap<String, String> parseRequestBody(String body) {
```
Comment (tbaik) [inline src/main/java/webserver/request/Parser.java]: I think this is for JSON bodies, right? we should probably state that it's only for JSON bodies!

Comment (tbaik) [reply src/main/java/webserver/request/Parser.java]: Could we also add tests for this?

File: src/main/java/webserver/JsonTodos.java [hunk 5]
```diff
@@ -14,29 +12,29 @@
 @SuppressWarnings("unchecked")
 public class JsonTodos implements Todos {
```
Comment (kkrull) [inline src/main/java/webserver/JsonTodos.java]: I didn't catch this earlier, but this looks like a good application of introducing an interface (or base class).  Now you can get TODO items from JSON files or just about anywhere else.

Take a moment and notice a few properties about the interface and its subclasses

* Adding additional implementations doesn't require the interface to change.  For this to happen, the methods, parameters, and return values in the interface methods have to be abstract enough to survive the addition of the next implementation class.  You're returning things like `JSONObject` (a class you wrote) and take in things like `int` IDs.  Those are pretty good abstractions -- give me an ID; I'll give you a TODO item.
* Different implementations do need different things.  You can't change the interface, so this information has to be provided in the constructors of the implementations.  Hence the `todosFile` constructor argument, which I suspect is not there in other implementations.
* TL;DR of all that: a stable interface deals strictly in the domain (using primitives or defining its own types as necessary), while implementations often differ in their constructor arguments.

The only thing I'd consider changing later would be to have a domain-centric notion of a TODO item (you use `JSONObject` and `JSONArray` so far in the types returned by these methods, which implies reliance on JSON).  So later on you might have a model class called `TODOItem` that may (or may not) be backed by a JSON data structure.  `JSONTODOItem` might then extend that class (or implement that interface).  Then this interface could deal in IDs and TODO items and stand the test of time.

File: src/main/java/webserver/JsonTodos.java [hunk 6]
```diff
@@ -50,19 +48,24 @@ public void addTodo(JSONObject newTodo) {
...
     }
 
-    public void updateTodo(JSONObject updatedTodo) {
+    public void updateTodo(int id, boolean isDone) {
         JSONArray todos = getAllTodos();
-        todos.add(updatedTodo);
-
-        try (FileWriter file = new FileWriter(todosFile)) {
+        try (FileWriter file = fileUtils.fileWriter(todosFile)) {
+            for (JSONObject todo: (Iterable<JSONObject>) todos) {
```
Comment (kkrull) [inline src/main/java/webserver/JsonTodos.java]: I've seen this pattern a few times, where you have to cast.  Usually (but not always), this can be avoided, along with the possible consequence of getting a `ClassCastException`.  

A good stepping stone might be to add a method to `JSONArray` that does the casting for you.

```java
class JSONArray {
  ...
  public Iterable<JSONObject> asIterable() {
    //KDK: totally guessing here
    return this.todoItems.stream()
      .map(x -> JSONObject.class.cast(x))
      .collect(Collectors.toList());
  }
  ...
}
```

File: src/main/java/webserver/Response.java [hunk 7]
```diff
@@ -55,7 +55,7 @@ private String createInitialResponseLine(int statusCode) {
     public String toResponseString() {
         String initialLine = createInitialResponseLine(statusCode);
         String allHeaders = convertHeaderMapToString(headers);
-        String responseString = initialLine + CRLF + allHeaders + CRLF + CRLF + body + "\n";
+        String responseString = initialLine + CRLF + allHeaders + CRLF + CRLF + body + CRLF + "\n";
```
Comment (kkrull) [inline src/main/java/webserver/Response.java]: Is the CRLF or the \n necessary here?  I'm not seeing it in the spec

https://tools.ietf.org/html/rfc2616#page-39

^^ note how the last thing in `response` is `[message-body]`; it doesn't say you have to have any special ending to the body.

If you're setting `content-length` in the response (which is helpful, but not required), I wonder if the `content-length` header is 3 bytes smaller than the content?  This might lead to things like `curl` giving warnings, or possibly to other clients truncating extra whitespace off the end of the response.

File: src/main/java/webserver/ResponseBody.java [hunk 8]
```diff
@@ -23,27 +23,25 @@ public static String getHtml(String path) throws IOException {
...
+          return new FileInputStream(path.toString());
       } catch (FileNotFoundException e) {
           Path errorFile = createFilePath("/404.html");
           return getFile(errorFile);
       }
   }
 
   private static String createHTMLString(FileInputStream file) throws IOException {
-      StringBuffer buf = new StringBuffer();
+      StringBuilder buf = new StringBuilder();
```
[the remark this answers is not in the document]
Comment (kkrull) [reply src/main/java/webserver/ResponseBody.java]: Agreed; you might say what it contains rather than what it is.  i.e. call it `html`

that's based on my pet peeve of saying things like `StringBuilder builder = new StringBuilder` -- where you say the word "builder" 3 times.

File: src/main/java/webserver/Router.java [hunk 9]
```diff
@@ -36,21 +45,22 @@ public String route(Request request) {
         assert routes != null;
         Optional<Callback<Request, String>> controller = routes.entrySet()
                                                   .stream()
-                                                  .filter( entry -> routeMatcher(requestPath, entry.getKey()))
+                                                  .filter(entry -> routeMatcher(requestPath, entry.getKey()))
```
Comment (kkrull) [inline src/main/java/webserver/Router.java]: :tada: Streaming!  Old school Java programmers lose their minds when they see this, even though Java 8 has been out for over 5 years now.  Now that you've done Clojure, you may proceed fearless into the use of the more functional style.

File: src/main/java/webserver/Router.java [hunk 10]
```diff
@@ -10,19 +10,28 @@
 import webserver.request.Request;
 import webserver.request.HTTP;;
 
-
 public class Router {
     private HashMap<String, Callback<Request, String>> GETRoutes = new HashMap<>();
+    private HashMap<String, Callback<Request, String>> HEADRoutes = new HashMap<>();
```
Comment (kkrull) [inline src/main/java/webserver/Router.java]: Here's a place where it's generally recommended to use a more generic interface as the declared type (and why there's all that repetition in declaring types for both the field and its value).

"Require no more (than you need); promise no less (than you will deliver)" is the general advice.

http://www.javapractices.com/topic/TopicAction.do?Id=194


TL;DR saying `Map<String, Callback...> headRoutes = new HashMap<>();` allows you to eventually accept types like `TreeMap` or `Hashtable` if those specific implementations are needed.

File: src/main/resources/todos.mustache [hunk 11]
```diff
@@ -36,4 +36,29 @@
             &copy 2019 @angieg0nzalez | 8th Light Apprenticeship
         </footer>
     </body>
+    <script>
+        document
```
Comment (kkrull) [inline src/main/resources/todos.mustache]: I wouldn't hold up this PR since the focus is on the Java code.  However I think it's possible this could try to look up the `#todo-form` element and add an event listener before the DOM is finished loading.  You might try wrapping all this code in 

```javascript
document.addEventListener("DOMContentLoaded", function(event) { 
  //do work
});
```

https://stackoverflow.com/questions/799981/document-ready-equivalent-without-jquery

File: src/main/java/webserver/SocketHandler.java [hunk 12]
```diff
@@ -27,18 +26,14 @@ public SocketHandler(Socket client, String directory) throws  IOException {
 
     public void run() {
         Parser parser= new Parser();
```
[the remark this answers is not in the document]
Comment (agonzalez0515) [reply src/main/java/webserver/SocketHandler.java]: 🙈

File: src/main/java/webserver/Router.java (hunk 10 above)
Comment (agonzalez0515) [reply src/main/java/webserver/Router.java]: Oh gotcha! Thanks!

File: src/main/java/webserver/Router.java (hunk 9 above)
Comment (agonzalez0515) [reply src/main/java/webserver/Router.java]: This was fun to look up! I saw it on a blogpost and took a bit to walk through but I really liked it.
